rpcs3/Emu/RSX/Common/surface_utils.h

```diff
@@ -178,51 +180,64 @@ namespace rsx
 		virtual bool is_depth_surface() const = 0;
 		virtual void release_ref(image_storage_type) const = 0;
 
-		virtual u32 get_surface_width(rsx::surface_metrics metrics = rsx::surface_metrics::pixels) const
+		template<rsx::surface_metrics Metrics = rsx::surface_metrics::pixels>
+		u32 get_surface_width() const
 		{
-			switch (metrics)
+			if constexpr (Metrics == rsx::surface_metrics::samples)
 			{
-				case rsx::surface_metrics::samples:
-					return surface_width * samples_x;
-				case rsx::surface_metrics::pixels:
-					return surface_width;
-				case rsx::surface_metrics::bytes:
-					return native_pitch;
-				default:
-					fmt::throw_exception("Unknown surface metric %d", u32(metrics));
+				return surface_width * samples_x;
+			}
+			else if constexpr (Metrics == rsx::surface_metrics::pixels)
+			{
+				return surface_width;
+			}
+			else if constexpr (Metrics == rsx::surface_metrics::bytes)
+			{
+				return native_pitch;
+			}
+			else
+			{
+				fmt::throw_exception("Unreachable");
```

Comment:
Wouldn't these have to be static asserts now?

Reply:
Clang/gcc disagree. I had static assert there before. Apparently its against spec to have static assert in a template without referencing T. There are workarounds but require adding statements outside the method and I did not like that.